File: tests/test_vcard.py
"""
Tests for vCard generation functionality.
"""
import pytest
from app.vcard import generate_vcard, generate_vcard_filename


def test_generate_vcard_minimal():
    """Test vCard generation with only required fields."""
    vcard = generate_vcard(name="John Doe")
    
    assert "BEGIN:VCARD" in vcard
    assert "VERSION:3.0" in vcard
    assert "FN:John Doe" in vcard
    assert "END:VCARD" in vcard


def test_generate_vcard_complete():
    """Test vCard generation with all fields."""
    vcard = generate_vcard(
        name="Jane Smith",
        company="Acme Corp",
        title="Software Engineer",
        email="jane@acme.com",
        phone="+1-555-123-4567",
        website="https://acme.com"
    )
    
    assert "FN:Jane Smith" in vcard
    assert "ORG:Acme Corp" in vcard
    assert "TITLE:Software Engineer" in vcard
    assert "EMAIL:jane@acme.com" in vcard
    assert "TEL:+1-555-123-4567" in vcard
    assert "URL:https://acme.com" in vcard


def test_generate_vcard_website_protocol():
    """Test that website gets https protocol if missing."""
    vcard = generate_vcard(
        name="Test User",
        website="example.com"
    )
    
    assert "URL:https://example.com" in vcard


def test_generate_vcard_phone_cleaning():
    """Test phone number cleaning."""
    vcard = generate_vcard(
        name="Test User",
        phone="(555) 123-4567"
    )
    
    assert "TEL:5551234567" in vcard


def test_generate_vcard_filename():
    """Test filename generation."""
    filename = generate_vcard_filename("John Doe")
    assert filename == "john-doe.vcf"
    
    filename = generate_vcard_filename("Jane O'Connor-Smith")
    assert filename == "jane-o-connor-smith.vcf"
    
    filename = generate_vcard_filename("")
    assert filename == "contact.vcf"

File: tests/test_vcard.js
import { generateVCard, generateVCardFilename } from '../server/utils/vcard.js'

describe('vCard Generation', () => {
  test('should generate valid vCard with all fields', () => {
    const contact = {
      name: 'John Doe',
      company: 'Acme Corp',
      title: 'Software Engineer',
      email: 'john@acme.com',
      phone: '+1-555-123-4567',
      website: 'https://acme.com'
    }

    const vcard = generateVCard(contact)
    
    expect(vcard).toContain('BEGIN:VCARD')
    expect(vcard).toContain('VERSION:3.0')
    expect(vcard).toContain('FN:John Doe')
    expect(vcard).toContain('ORG:Acme Corp')
    expect(vcard).toContain('TITLE:Software Engineer')
    expect(vcard).toContain('EMAIL:john@acme.com')
    expect(vcard).toContain('TEL:+15551234567')
    expect(vcard).toContain('URL:https://acme.com')
    expect(vcard).toContain('END:VCARD')
  })

  test('should generate vCard with only required fields', () => {
    const contact = {
      name: 'Jane Smith'
    }

    const vcard = generateVCard(contact)
    
    expect(vcard).toContain('BEGIN:VCARD')
    expect(vcard).toContain('VERSION:3.0')
    expect(vcard).toContain('FN:Jane Smith')
    expect(vcard).toContain('END:VCARD')
    expect(vcard).not.toContain('ORG:')
    expect(vcard).not.toContain('EMAIL:')
  })

  test('should throw error for missing name', () => {
    const contact = {
      company: 'Acme Corp'
    }

    expect(() => generateVCard(contact)).toThrow('Name is required')
  })

  test('should generate valid filename', () => {
    expect(generateVCardFilename('John Doe')).toBe('john-doe.vcf')
    expect(generateVCardFilename('Jane Smith-Jones')).toBe('jane-smith-jones.vcf')
    expect(generateVCardFilename('')).toBe('contact.vcf')
    expect(generateVCardFilename('   ')).toBe('contact.vcf')
  })

  test('should clean phone numbers', () => {
    const contact = {
      name: 'Test User',
      phone: '+1 (555) 123-4567'
    }

    const vcard = generateVCard(contact)
    expect(vcard).toContain('TEL:+15551234567')
  })

  test('should add protocol to website URLs', () => {
    const contact = {
      name: 'Test User',
      website: 'example.com'
    }

    const vcard = generateVCard(contact)
    expect(vcard).toContain('URL:https://example.com')
  })
})
